pkg/csi_driver/node.go

```diff
@@ -450,3 +477,122 @@ func getFSStat(path string) (available, capacity, used, inodesFree, inodes, inod
+		return nil
+	}
+
+	_, err = util.UpdateConfigMapWithKeyValue(ctx, cm, lockInfoKey, filestoreIP, s.kubeClient)
+	if err != nil {
+		klog.Errorf("NodeStageVolume failed to update configmap %s/%s with lock info {%s: %s} for volume %s: %v", util.ConfigMapNamespace, configmapName, volumeID, err)
+		return err
+	}
+
+	return nil
+}
+
+// nodeUnstageVolumeUpdateLockInfo updates lock info after NodeUnStageVolume succeed.
+func (s *nodeServer) nodeUnstageVolumeUpdateLockInfo(ctx context.Context, req *csi.NodeUnstageVolumeRequest) error {
+	volumeID := req.GetVolumeId()
+	// No-op if lock release is not supported by the driver.
+	if !s.features.FeatureLockRelease {
+		klog.Infof("Skip lock release info update for volume %s: feature lock release not enabled", volumeID)
+		return nil
+	}
+
+	stagingTargetPath := req.StagingTargetPath
+	klog.V(4).Infof("NodeUnstageVolume succeeded on volume %v from staging target path %s, proceed to lock info configmap updates", volumeID, stagingTargetPath)
+	nodeName := s.driver.config.NodeName
+	configmapName := util.ConfigMapNamePrefix + nodeName
+	klog.Infof("NodeUnstageVolume getting configmap %s/%s for volume %s", util.ConfigMapNamespace, configmapName, volumeID)
+	cm, err := util.GetConfigMap(ctx, configmapName, util.ConfigMapNamespace, s.kubeClient)
+	if err != nil {
+		klog.Errorf("NodeStageVolume failed to get configmap %s/%s for volume %s: %v", util.ConfigMapNamespace, configmapName, volumeID, err)
+		return err
+	}
+	if cm == nil {
+		klog.Infof("NodeUnstageVolume skipped updating lock info for volume %s: configmap %s/%s not found", volumeID, util.ConfigMapNamespace, configmapName)
+		return nil
+	}
+
+	lockInfoKey, err := s.generateLockInfoKeyFromVolumeID(volumeID)
+	if err != nil {
+		klog.Errorf("NodeUnstageVolume failed to generate lock info key for volume %s: %v", volumeID, err)
+		return err
+	}
+
+	if _, err := util.RemoveKeyFromConfigMap(ctx, cm, lockInfoKey, s.kubeClient); err != nil {
+		klog.Infof("NodeUnstageVolume failed to remove key %s from configmap %s/%s for volume %s: %v", lockInfoKey, cm.Namespace, cm.Name, volumeID, err)
+		return err
+	}
+
+	return nil
+}
+
+// generateLockInfoKeyFromVolumeID generates a configmap key for the given volumeID.
+// The configmap will store key-value pairs in format:
+// {projectID}.{location}.{filestoreName}.{shareName}.{gkeNodeID}.{gkeNodeInternalIP}: <filestoreIP>
+func (s *nodeServer) generateLockInfoKeyFromVolumeID(volumeID string) (string, error) {
+	var lockInfoKey string
+	nodeID := s.metaService.GetInstanceID()
+	nodeInternalIP := s.metaService.GetInternalIP()
+	if isMultishareVolId(volumeID) {
+		_, project, location, filestoreName, shareName, err := parseMultishareVolId(volumeID)
+		if err != nil {
+			return "", err
+		}
+		lockInfoKey = util.GenerateConfigMapKey(project, location, filestoreName, shareName, nodeID, nodeInternalIP)
+	} else {
```

Comment:
nit: we can drop the else. return from the if above

Reply:
done